core/state/database.go

```diff
@@ -176,3 +239,151 @@ func (m cachedTrie) Commit(onleaf trie.LeafCallback) (common.Hash, error) {
 func (m cachedTrie) Prove(key []byte, fromLevel uint, proofDb ethdb.Putter) error {
 	return m.SecureTrie.Prove(key, fromLevel, proofDb)
 }
+
+// syncedTrie is a synchronized wrapper around a trie to support multiple threads
+// accessing and read-expanding the same trie (i.e. caching loaded trie nodes).
+type syncedTrie struct {
+	tr   *trie.SecureTrie
+	lock sync.Mutex
```

Comment:
Any reason not to use a `sync.RWMutex` here instead?  If there's a strong argument for it being a traditional mutex perhaps a comment should be added so others don't try to update it to a RWMutex later?

Reply:
A trie gets modified if a read expands a new path as it needs to load data from disk. As such, any operation on the trie needs to be exclusive.